ArduCopter/mode_rc_car.cpp

```diff
@@ -0,0 +1,1000 @@
+        if (copter.ap.land_complete_maybe) {
+            loiter_nav->soften_for_landing();
+        }
+
+        // Pos Hold State Machine Determination
+        AltHoldModeState poshold_state = get_alt_hold_state(target_climb_rate);
+
+        // state machine
+        switch (poshold_state) {
+
+        case AltHold_MotorStopped:
+            attitude_control->reset_rate_controller_I_terms();
+            attitude_control->reset_yaw_target_and_rate(false);
+            pos_control->relax_z_controller(0.0f);   // forces throttle output to decay to zero
+            loiter_nav->clear_pilot_desired_acceleration();
+            loiter_nav->init_target();
+
+            // set poshold state to pilot override
+            roll_mode = RPMode::PILOT_OVERRIDE;
+            pitch_mode = RPMode::PILOT_OVERRIDE;
+
+            // initialise wind compensation estimate
+            init_wind_comp_estimate();
+            break;
+
+        case AltHold_Landed_Ground_Idle:
+            loiter_nav->clear_pilot_desired_acceleration();
+            loiter_nav->init_target();
+            attitude_control->reset_yaw_target_and_rate();
+            init_wind_comp_estimate();
+            FALLTHROUGH;
+
+        case AltHold_Landed_Pre_Takeoff:
+            attitude_control->reset_rate_controller_I_terms_smoothly();
+            pos_control->relax_z_controller(0.0f);   // forces throttle output to decay to zero
+
+            // set poshold state to pilot override
+            roll_mode = RPMode::PILOT_OVERRIDE;
+            pitch_mode = RPMode::PILOT_OVERRIDE;
+            break;
+
+        case AltHold_Takeoff:
+            // initiate take-off
+            if (!takeoff.running()) {
+                takeoff.start(constrain_float(g.pilot_takeoff_alt,0.0f,1000.0f));
+            }
+
+            // get avoidance adjusted climb rate
+            target_climb_rate = get_avoidance_adjusted_climbrate(target_climb_rate);
+
+            // set position controller targets adjusted for pilot input
+            takeoff.do_pilot_takeoff(target_climb_rate);
+
+            // init and update loiter although pilot is controlling lean angles
+            loiter_nav->clear_pilot_desired_acceleration();
+            loiter_nav->init_target();
+
+            // set poshold state to pilot override
+            roll_mode = RPMode::PILOT_OVERRIDE;
+            pitch_mode = RPMode::PILOT_OVERRIDE;
+            break;
+
+        case AltHold_Flying:
+            motors->set_desired_spool_state(AP_Motors::DesiredSpoolState::THROTTLE_UNLIMITED);
+
+            // get avoidance adjusted climb rate
+            target_climb_rate = get_avoidance_adjusted_climbrate(target_climb_rate);
+
+            // update the vertical offset based on the surface measurement
+            copter.surface_tracking.update_surface_offset();
+
+            // Send the commanded climb rate to the position controller
+            if (copter.g.rccar_unlimited_height == 0) {	
```

Comment:
Use absolute z position when in mode 1 and climb rate when in mode 2. There may be weirdness with the way I've done this, since this can be switched mid flight